packages/react-native-reanimated/plugin/src/class.ts

```diff
@@ -0,0 +1,331 @@
+  variableDeclarator,
+} from '@babel/types';
+import type {
+  BlockStatement,
+  ClassDeclaration,
+  Expression,
+  FunctionExpression,
+  Program,
+  File as BabelFile,
+  VariableDeclaration,
+  Identifier,
+  ClassBody,
+  ExportDefaultDeclaration,
+  ExportNamedDeclaration,
+} from '@babel/types';
+import type { ReanimatedPluginPass } from './types';
+import generate from '@babel/generator';
+import { WorkletizableFunction } from './types';
+import traverse from '@babel/traverse';
+import { strict as assert } from 'assert';
+
+const classWorkletMarker = '__workletClass';
+
+export function processIfWorkletClass(
+  classPath: NodePath<ClassDeclaration>,
+  state: ReanimatedPluginPass
+): boolean {
+  if (!classPath.node.id) {
+    // We don't support unnamed classes yet.
+    return false;
+  }
+
+  if (!hasWorkletClassMarker(classPath.node.body)) {
+    return false;
+  }
+
+  const parentPath = classPath.parentPath;
+  const className = classPath.node.id.name;
+
+  classPath.skip();
+
+  classPath = makeStatement(classPath, parentPath, className);
+
+  removeWorkletClassMarker(classPath.node.body);
+
+  processClass(classPath, state);
+
+  return true;
+}
+
+function processClass(
+  classPath: NodePath<ClassDeclaration>,
+  state: ReanimatedPluginPass
+) {
+  assert(classPath.node.id);
+  const className = classPath.node.id.name;
+  const code = generate(classPath.node).code;
+
+  const transformedCode = transformSync(code, {
+    plugins: [
+      '@babel/plugin-transform-class-properties',
+      '@babel/plugin-transform-classes',
+      '@babel/plugin-transform-unicode-regex',
+    ],
+    filename: state.file.opts.filename,
+    ast: true,
+    babelrc: false,
+    configFile: false,
+  });
+
+  assert(transformedCode);
+  assert(transformedCode.ast);
+
+  const ast = transformedCode.ast;
+
+  let factory: VariableDeclaration;
+
+  let hasHandledClass = false;
+
+  traverse(ast, {
+    [WorkletizableFunction]: {
+      // @ts-expect-error TS has some trouble inferring here.
+      enter: (functionPath: NodePath<WorkletizableFunction>) => {
+        if (functionPath.parentPath.isObjectProperty()) {
+          return;
+        }
+
+        const workletDirective = directive(directiveLiteral('worklet'));
+
+        if (functionPath.parentPath.isCallExpression() && !hasHandledClass) {
+          const factoryCopy = cloneNode(
+            functionPath.node,
+            true
+          ) as FunctionExpression;
+          factoryCopy.id = identifier(className + 'ClassFactory');
+          factoryCopy.body.directives.push(workletDirective);
+          factory = variableDeclaration('const', [
+            variableDeclarator(
+              identifier(className + 'ClassFactory'),
+              factoryCopy
+            ),
+          ]);
+          hasHandledClass = true;
+
+          return;
+        }
+
+        const bodyPath = functionPath.get('body');
+        if (!bodyPath.isBlockStatement()) {
+          bodyPath.replaceWith(
```

Comment:
this can be extracted to a utility function

Reply:
I don't need it anymore.